app/questionnaire/navigator.py

```diff
@@ -1,193 +1,286 @@
+        repeat_index = repeat_rule['answer_id']
+        filtered = answers.filter(answer_id=repeat_index)
+        if repeat_rule['type'] == 'answer_value':
+            if len(filtered) == 1:
+                return int(filtered[0]['value'])
+            return 1
+        elif repeat_rule['type'] == 'answer_count':
+            return len(filtered)
 
 
 class Navigator:
     PRECEEDING_INTERSTITIAL_PATH = ['introduction']
     CLOSING_INTERSTITIAL_PATH = ['summary', 'thank-you']
 
-    def __init__(self, survey_json):
+    def __init__(self, survey_json, answer_store=None):
+        self.answer_store = answer_store or AnswerStore()
         self.survey_json = survey_json
-        self.blocks = self.get_blocks()
-        self.first_block_id = self.blocks[0]['id']
+        self.first_block_id = self.get_first_block_id()
+        self.first_group_id = self.get_first_group_id()
+        self.last_group_id = self.get_last_group_id()
+        self.location_path = self.get_location_path()
 
-    def get_routing_path(self, answers=None):
+    def build_path(self, blocks, group_id, group_instance, block_id, path):
         """
-        Returns a list of the block ids visited based on answers provided
-        :return: List of block ids
+        Recursive method which visits all the blocks and returns path taken
+        given a list of answers
+
+        :param blocks: A list containing all blocks in the survey
+        :param group_id: The group id to visit
+        :param group_instance: The group instance to visit
+        :param block_id: The block id to visit
+        :param path: The known path as a list which has been visited already
+        :return: A list of block ids followed through the survey
         """
-        answers = answers or {}
-        routing_path = []
+        if block_id in self.CLOSING_INTERSTITIAL_PATH:
+            return path
+
+        this_block = {
+            "block_id": block_id,
+            "group_id": group_id,
+            "group_instance": group_instance,
+        }
+
+        path.append(this_block)
+
+        # Return the index of the block id to be visited
+        block_id_index = next(index for (index, b) in enumerate(blocks) if b["block"]["id"] == block_id and
+                              b["group_id"] == group_id and b['group_instance'] == group_instance)
+
+        block = blocks[block_id_index]["block"]
+
+        if 'routing_rules' in block and len(block['routing_rules']) > 0:
+            for rule in block['routing_rules']:
+                if 'goto' in rule:
+                    should_go = evaluate_goto(rule['goto'], self.answer_store, group_instance)
+                    if should_go is True:
+                        return self.build_path(blocks, group_id, group_instance, rule['goto']['id'], path)
+                    elif should_go is False:
+                        return path
+        # If this isn't the last block in the set evaluated
```

Comment:
Alternatively to adding a comment we could have refactored to a variable that won't go stale!

```
last_block = block_id_index != len(blocks) - 1
elif not last_block:
```